Update hashes on older models

ljos opened the pull request:
- hunpos/hmm_tagger.ml (load): Check if we can find `<s>` in tag_vocab, if
  not we run the fix and return a model with an updated vocabulary.

- util/vocab.ml (fix_old_vocab): New function. Creates a new vocab where
  the ids and word are rehashed.

This fix makes it possible to use old models like [en_wsj.model](https://storage.googleapis.com/google-code-archive-downloads/v2/code.google.com/hunpos/en_wsj.model.gz) again without having to re-train the model.

ljos committed:
Fix #21: Update hashes on older models

* hunpos/hmm_tagger.ml (load): Check if we can find <s> in tag_vocab, if
  not we run the fix and return a model with an updated vocabulary.

* util/vocab.ml (fix_old_vocab): New function. Creates a new vocab where
  the ids and word are rehashed.

ljos commented:
@giuliopaci I see you are the one doing most of the changes, but you are not watching the repository. I thought I would just ping you to see if you agree to the changes and want to merge them.

giuliopaci commented:
Hi @ljos, nice to see that you are also working on hunpos. Thank you for the ping... Indeed I missed these pull requests. I am going to review them now. :-)

giuliopaci commented:
The code seems fine to me. I will try to run a test and then accept or request an update.

Can you remove `.gitignore` from this pull request? I do not like `.gitignore` files that hide important files in a repository. I think it is error prone and an incentive bad behaviors like `git add *`.

ljos commented:
I have just had a tendency to add build files to the repo and therefore I like .gitignore files.

ljos commented:
I see I don`t use the `max` variable in `fix_old_vocab`. I will fix that as well.

giuliopaci commented:
I understand your issue. I think this is a consequence of bad behaviors, like `git add *`... I also had that tendency when I was using `svn` and so I added a lot of ignores as well.
Unfortunately I have seen many more issues coming from that kind of ignored files later (e.g., having old builds in a "clean" working directory hiding bugs or leading to unexpected behaviors).

Maybe you may try something like what is suggested in http://stackoverflow.com/questions/1753070/git-ignore-files-only-locally if you want to ignore those files on your repository.

giuliopaci commented:
Let me know when I can merge the pull request. I have tested it a little bit and everything worked on my side.

ljos commented:
You can merge it now if you want.

ljos commented:
I would usually either use a proper clean command from a Makefile or build script or use `git clean -f` if I wanted to make sure all build files are removed; I can't remember having problems with bugs because of old builds.

However, I will stick to your conventions. It doesn't really bother me either way.

ljos commented:
Also, they mentioned in dkpro/dkpro-core#619 that they would like that we cut a new version and build binaries for the them to download. I don't know if that is something you would want to do; I don't even have a windows machine to test on.

giuliopaci commented:
I do not have a Windows machine either. And as far as I know github.com does not offer the possibility to upload prebuilt files (do you have different information?), so it will be difficult for me to help on this side, but not against it in general. :-)

ljos commented:
Github does support it now. If you go to the [new release page](https://github.com/mivoq/hunpos/releases/new) there is a field that says "Attach binaries". I can't go to that page on this repo, but I can see it on the ones where I am a contributor.

giuliopaci commented:
Great to know, thank you! :-) So apparently the only missing bits now are the binaries. :-)

reckart commented:
From DKPro Core perspective, even just having a new Linux binary would be helpful. My understanding is that there was only a minor modification to the code which shouldn't impact the actual results. So we could continue to use the old binaries for Windows and OS X and just use a new binary on Linux.

ljos commented:
@reckart The bug that I fixed was not because of 64 bit systems. It was because in v4.00.0 of Ocaml they changed the hash function for their hash table. The way the hunpos models are stored is that that they are trained and then serialized directly from memory to a binary file. That means it contains the hashes from the old hash function which will not work with the new hash function.

If on a 64 bit system you install the 32bit glib headers you shouldn't see any of the bugs you experienced. Also, if you compile hunpos and train a new model there is no problem. The problem I fixed was when you had a newly compiled hunpos and an old model. 

If you take the latest v3 of Ocaml instead of v4 I suspect you would be able to compile a new 64bit version as well without any problems.

ljos commented:
Also, I can always build binaries for all of the platforms, but I have no way of testing on windows. If someone can confirm that a binary is working there I think we can push a new release.

reckart commented:
I can test things on Windows.

ljos commented:
@reckart cool. I will build something and send it to you.
